### PR TITLE
Don't use `skopeo inspect` for canonical (digested) pull specs

### PR DESCRIPTION
**Pull request by cgwalters**
`skopeo inspect` hits the `/tags` API which can be expensive
for registries to compute.  Quay.io recently changed how
they do rate limiting, and we started hitting this in OpenShift CI
jobs.

Since the machine-config-operator always provides canonical/digested
pull specs, the non-digested support is just for local developer
convenience.

And anyways, what we really should do is pull the image unconditionally
and then inspect it afterwards, we don't need all of the tag info.

Basically this PR ends up in a place where the non-canonical
path only loses out on the optimization of avoiding pulling the
image in the case where we already have that target.

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -268,31 +269,44 @@ func pullAndRebase(container string) (imgid string, changed bool) {
...
+		args = append(args, container)
+		utils.RunExt(false, numRetriesNetCommands, "podman", args...)
 	}
 
-	// Pull the image
-	podmanArgs := []string{"pull", "-q"}
-	podmanArgs = append(podmanArgs, authArgs...)
-	podmanArgs = append(podmanArgs, imgid)
-	utils.RunExt(false, numRetriesNetCommands, "podman", podmanArgs...)
+	inspectArgs := []string{"inspect", "--type=image"}
```
Comment (miabbott) [inline cmd/root.go]: Shouldn't we continue to append the `authArgs` here?

**Comment by wking** [comment]
Is there a double-negative issue in the PR/commit subject?  I think we want to skip `skopeo inspect` for *canonical* pull-specs, not non-canonical pull-specs.

File: cmd/root.go (hunk 1 above)
Comment (miabbott) [reply cmd/root.go]: Ah, nm...I see that we are inspecting after it has been pulled.

**Comment by wking** [comment]
Is there somewhere we can/should document the fact that with this PR `pivot` no longer uses `skopeo` at all?

File: cmd/root.go [hunk 2]
```diff
@@ -268,31 +269,44 @@ func pullAndRebase(container string) (imgid string, changed bool) {
...
+	isCanonicalForm := true
+	if _, err := getRefDigest(container); err != nil {
+		isCanonicalForm = false
+		// In non-canonical form, we pull unconditionally right now
+		args := []string{"pull", "-q"}
+		args = append(args, authArgs...)
+		args = append(args, container)
+		utils.RunExt(false, numRetriesNetCommands, "podman", args...)
+	} else {
+		targetMatched, err := compareOSImageURL(previousPivot, container)
```
Comment (jlebon) [inline cmd/root.go]: Hmm, I see the reasoning in the commit message, though it wouldn't be hard to just adjust the logic so we still do this in the developer case too right? I.e.

```
if !canonical {
  podman pull
  podman inspect
  imgid = ...
} else {
  imgid = container
}
```

?

Then later on, `podman pull` by imgid, so in the developer case it's just a no-op.

Comment (jlebon) [reply cmd/root.go]: IOW, I'm not sure if

> Basically this PR ends up in a place where the non-canonical
path only loses out on the optimization of avoiding pulling the
image in the case where we already have that target.

is true since the `targetMatched` logic isn't done unconditionally, right? I.e. we can hit the `Old and new refs are equal` error.

Comment (cgwalters) [reply cmd/root.go]: > i.e. we can hit the Old and new refs are equal error.

That's [not an error anymore though](https://github.com/projectatomic/rpm-ostree/pull/1807)

**Comment by cgwalters** [comment]
> Is there somewhere we can/should document the fact that with this PR pivot no longer uses skopeo at all?

Not sure why that'd matter?  The few people who should care are all reading this PR.

> though it wouldn't be hard to just adjust the logic so we still do this in the developer case too right? I.e.

Since as noted above this isn't a hard error anymore, all we're avoiding is re-pulling the image and *only* when the sha256 would match, and if it does, why are you asking `pivot` to pull it?  Also I tested this PR manually in a few scenarios...fixes a pretty important issue, I don't think it's worth redoing just for that.

**Comment by jlebon** [comment]
Ahh right, forgot about https://github.com/projectatomic/rpm-ostree/pull/1807.

/lgtm

**Comment by wking** [comment]
Commit message is fixed, but PR subject still has the double-negative.
